src/parse.jl

```diff
@@ -0,0 +1,361 @@
+  * The `percentages` field is a dictionary with keys of type `Percent`, which is a custom type. The `liftkey` function is defined to convert the JSON string keys to `Percent` types (parses the Float64 manually)
+  * The `json_properties` field has a type of `JSONText`, which means the raw JSON will be preserved as a String of the `JSONText` type.
+  * The `matrix` field is a `Matrix{Float64}`, so the JSON input array-of-arrays are materialized as such.
+  * The `extra_key` field is not defined in the `FrankenStruct` type, so it is ignored and skipped over.
+
+NOTE:
+Why use `JSON.Object{String, Any}` as the default object type? It provides several benefits:
+  * Behaves as a drop-in replacement for `Dict{String, Any}`, so no loss of functionality
+  * Performance! It's internal representation means memory savings and faster construction for small objects typical in JSON (vs `Dict`)
+  * Insertion order is preserved, so the order of keys in the JSON input is preserved in `JSON.Object`
+  * Convenient `getproperty` (i.e. `obj.key`) syntax is supported, even for `Object{String,Any}` key types (again ideal/specialized for JSON usage)
+
+`JSON.Object` internal representation uses a linked list, thus key lookups are linear time (O(n)). For *large* JSON objects,
+(hundreds or thousands of keys), consider using a `Dict{String, Any}` instead, like `JSON.parse(json; dicttype=Dict{String, Any})`.
+"""
+function parse end
+
+import StructUtils: StructStyle
+
+abstract type JSONStyle <: StructStyle end
+
+# defining a custom style allows us to pass a non-default dicttype `O` through JSON.parse
+struct JSONReadStyle{O, T} <: JSONStyle
+    null::T
+end
+
+JSONReadStyle{O}(null::T) where {O, T} = JSONReadStyle{O, T}(null)
+
+objecttype(::StructStyle) = DEFAULT_OBJECT_TYPE
+objecttype(::JSONReadStyle{OT}) where {OT} = OT
+nullvalue(::StructStyle) = nothing
+nullvalue(st::JSONReadStyle) = st.null
+
+# this allows struct fields to specify tags under the json key specifically to override JSON behavior
+StructUtils.fieldtagkey(::JSONStyle) = :json
+
+function parsefile end
+@doc (@doc parse) parsefile
+
+function parsefile! end
+@doc (@doc parse) parsefile!
+
+parsefile(file; jsonlines::Union{Bool, Nothing}=nothing, kw...) = open(io -> parse(io; jsonlines=(jsonlines === nothing ? isjsonl(file) : jsonlines), kw...), file)
+parsefile(file, ::Type{T}; jsonlines::Union{Bool, Nothing}=nothing, kw...) where {T} = open(io -> parse(io, T; jsonlines=(jsonlines === nothing ? isjsonl(file) : jsonlines), kw...), file)
+parsefile!(file, x::T; jsonlines::Union{Bool, Nothing}=nothing, kw...) where {T} = open(io -> parse!(io, x; jsonlines=(jsonlines === nothing ? isjsonl(file) : jsonlines), kw...), file)
+
+parse(io::Union{IO, Base.AbstractCmd}, ::Type{T}=Any; kw...) where {T} = parse(Base.read(io), T; kw...)
+
+function parse(io::IOStream, ::Type{T}=Any; kw...) where {T}
+    buf = Mmap.mmap(io)
+    res = parse(buf, T; kw...)
+    checkfinalize!(buf)
+    return res
+end
```

Comment:
I don't think `mmap` can be reliably supported here, because `checkfinalize!(buf)` currently needs to access private fields of `buf`.